triton-vm/src/table/constraint_circuit.rs

```diff
@@ -1060,6 +875,182 @@ impl<II: InputIndicator> ConstraintCircuitMonad<II> {
+
+            // Create a new variable.
+            let chosen_node = builder.get_node_by_id(chosen_node_id).unwrap();
+            let chosen_node_is_base_col = chosen_node.circuit.borrow().evaluates_to_base_element();
+            let new_col_idx = match chosen_node_is_base_col {
+                true => num_base_cols + base_constraints.len(),
+                false => num_ext_cols + ext_constraints.len(),
+            };
+            let new_input_indicator = match chosen_node_is_base_col {
+                true => II::base_table_input(new_col_idx),
+                false => II::ext_table_input(new_col_idx),
+            };
+            let new_variable = builder.input(new_input_indicator);
+            let new_circuit = new_variable.circuit.clone();
+
+            // Substitute the chosen circuit with the new variable.
+            builder.substitute(chosen_node_id, new_circuit.clone());
+
+            // Create new constraint and put it into the appropriate return vector.
+            let new_constraint = new_variable - chosen_node;
+            match chosen_node_is_base_col {
+                true => base_constraints.push(new_constraint),
+                false => ext_constraints.push(new_constraint),
+            }
+
+            // Treat roots of the multicircuit explicitly.
+            for circuit in multicircuit.iter_mut() {
+                if circuit.circuit.borrow().id == chosen_node_id {
+                    circuit.circuit = new_circuit.clone();
+                }
+            }
+        }
+
+        (base_constraints, ext_constraints)
+    }
+
+    /// Heuristically pick a node from the given multicircuit that is to be substituted with a new
+    /// variable. The ID of the chosen node is returned.
+    fn pick_node_to_substitute(
+        multicircuit: &[ConstraintCircuitMonad<II>],
+        target_degree: Degree,
+    ) -> usize {
+        if multicircuit.is_empty() {
+            panic!("Multicircuit must be non-empty in order to pick a node from it.");
+        }
+
+        let multicircuit = multicircuit
+            .iter()
+            .map(|c| c.clone().consume())
+            .collect_vec();
+        let all_nodes = Self::all_nodes_in_multicircuit(&multicircuit);
+        let all_nodes: HashSet<_> = HashSet::from_iter(all_nodes.into_iter());
+
+        // Only nodes with degree > target_degree need changing.
+        let high_degree_nodes = all_nodes
+            .into_iter()
+            .filter(|node| node.degree() > target_degree)
+            .collect_vec();
+
+        // Collect all the nodes where some substitution is necessary.
+        // Substituting a node of degree 1 is both pointless and can lead to infinite iteration.
+        let low_degree_nodes = Self::all_nodes_in_multicircuit(&high_degree_nodes)
+            .into_iter()
+            .filter(|node| node.degree() <= target_degree)
+            .filter(|node| node.degree() > 1)
+            .collect_vec();
+
```

Comment:
The variable name seems to be confusing. Here, I call `low_degree_nodes` whatever falls below the `target_degree`. While the substitution has to happen in a high degree node, the node that is substituted has to be of low degree (in the above sense), otherwise the newly generated constraint would itself be of high degree. This might lead to trouble with termination, and potentially create other kinds of problems.

I'm happy about suggestions for better variable names. Until then, I'll throw in a few more comments.